### PR TITLE
(PE-30464) Implement `project_inventory_targets` in bolt-server

### PR DESCRIPTION
**Pull request by donoghuc**
This commit adds a new `project_inventory_targets` endpoint to bolt-server that parses a project inventory and returns a list of hashes describing targets parsed from that inventory.

### CODE REVIEW

File: lib/bolt_server/transport_app.rb [hunk 1]
```diff
@@ -630,6 +634,28 @@ def make_winrm_target(target_hash)
...
+    get '/project_inventory_targets' do
+      bolt_config = config_from_project(params['project_ref'])
+      if bolt_config.inventoryfile && bolt_config.project.inventory_file.to_s != bolt_config.inventoryfile
+        raise Bolt::ValidationError, "Project inventory must be defined in #{bolt_config.project.inventory_file}"
+      end
+      plugin = Bolt::Plugin.setup(bolt_config, nil)
+      inventory = Bolt::Inventory.from_config(bolt_config, plugin)
+      target_list = inventory.get_targets('all').map do |target|
+        target_data = {
+          'name' => target.name,
```
Comment (donoghuc) [inline lib/bolt_server/transport_app.rb:649]: Oops, i guess `Target#to_h` already has `name`. Will remove.

**Review by nicklewis — CHANGES_REQUESTED** [review]
I think in general we want to avoid using the factory methods for many of these objects as those implement a lot of Bolt-specific logic. Ideally, we ought to be able to handle loading the data on our own and then just construct objects directly. That's just an observation though, not a request for changes on this PR in particular. We may need to make some improvements in Bolt to support it as well.

File: lib/bolt_server/transport_app.rb [hunk 2]
```diff
@@ -630,6 +634,28 @@ def make_winrm_target(target_hash)
...
 
+    # Returns a list of targets parsed from a Project inventory
+    #
+    # @param project_ref [String] the project_ref to compute the inventory from
+    get '/project_inventory_targets' do
+      bolt_config = config_from_project(params['project_ref'])
+      if bolt_config.inventoryfile && bolt_config.project.inventory_file.to_s != bolt_config.inventoryfile
+        raise Bolt::ValidationError, "Project inventory must be defined in #{bolt_config.project.inventory_file}"
+      end
+      plugin = Bolt::Plugin.setup(bolt_config, nil)
```
[the remark this answers is not in the document]
Comment (ekinanp) [reply lib/bolt_server/transport_app.rb:645]: Agreed. I don't think that needs to be in this PR though because we're the ones that will be using this endpoint, and we won't be using inventory files with plugins. We'll also be extending this with proper plugin support eventually before a release and part of that work does include handling the built-in plugins.

File: lib/bolt_server/transport_app.rb [hunk 3]
```diff
@@ -630,6 +634,28 @@ def make_winrm_target(target_hash)
...
+      inventory = Bolt::Inventory.from_config(bolt_config, plugin)
```
[the remark this answers is not in the document]
Comment (ekinanp) [reply lib/bolt_server/transport_app.rb:646]: Not sure I like that approach because it means we'll also need to remember to do https://github.com/puppetlabs/bolt/blob/main/lib/bolt/inventory.rb#L72-L75 when we get to Inventory plugins.

Re: "any plugins running now", I think that will still be an issue due to https://github.com/puppetlabs/bolt/blob/main/lib/bolt/inventory/group.rb#L21 but it is something we will eventually fix before releasing this so doesn't seem like a big deal.

I do agree that the `BOLT_INVENTORY` env. var bit makes `from_config` annoying to use. But I do like that when `BOLT_INVENTORY` is not set, `from_config` ensures I get an Inventory object constructed from my config that's ready for me to use (so I don't have to worry about e.g. resolving plugin references in the bolt config).

Since we won't be running into the `BOLT_INVENTORY` env. var issue in our current work, maybe we could keep `from_config` for now?

I do think it'd be worth adding a `Bolt::Inventory.from_config_strict(bolt_config, plugin)` method to Bolt that will load the inventory strictly from the config. That would satisfy our use-case and avoid the `BOLT_INVENTORY` environment variable issue.

File: spec/bolt_server/transport_app_spec.rb [hunk 4]
```diff
@@ -787,6 +795,44 @@ def post_over_transport(transport, action, body_content, multiple: false)
       end
     end
 
+    describe '/project_inventory_targets' do
+      let(:bolt_project) { { 'name' => 'my_project' } }
+      let(:targets) { %w[one two three] }
```
Comment (ekinanp) [inline spec/bolt_server/transport_app_spec.rb:800]: Would it be worth including Nick's edge-case for `protocol` as a test case here to make sure that `transport` properly handles that? Thinking his example of

```
targets:
  - uri: ssh://foo.example.com
    config:
      transport: remote
```

where `protocol` would be `ssh`, but `transport` is `remote`.

File: lib/bolt_server/transport_app.rb [hunk 5]
```diff
@@ -261,13 +261,17 @@ def in_pe_pal_env(environment)
       end
     end
 
-    def in_bolt_project(bolt_project)
-      return [400, '`project_ref` is a required argument'] if bolt_project.nil?
-      project_dir = File.join(@config['projects-dir'], bolt_project)
-      return [400, "`project_ref`: #{project_dir} does not exist"] unless Dir.exist?(project_dir)
+    def config_from_project(project_ref)
+      raise Bolt::ValidationError, '`project_ref` is a required argument' if project_ref.nil?
```
[the remark this answers is not in the document]
Comment (ekinanp) [reply lib/bolt_server/transport_app.rb:265]: Definitely separate from this PR.

File: lib/bolt_server/transport_app.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (McdonaldSeanp) [reply lib/bolt_server/transport_app.rb:646]: If there are other reasons to refactor this that seems fine.

But I don't think we need to worry about misconfiguration from user environment vars as a top concern. "I updated my replicated configuration so that the orchestrator container instance runs with the `BOLT_INVENTORY` env var set" isn't something I expect we'll find often. 

If a user _did_ start trying to pass env vars to the running container instances, the fix is to have them stop doing that.

[the remark this answers is not in the document]
Comment (nicklewis) [reply lib/bolt_server/transport_app.rb:646]: I don't think it's a huge deal. It's mostly reflective of a larger problem which is that there are cases where we can't currently avoid Bolt-specific logic that is potentially harmful or at least unneeded.

[the remark this answers is not in the document]
Comment (McdonaldSeanp) [reply lib/bolt_server/transport_app.rb:646]: That's fair, if our argument is for better code maintainability and less coupling to bolt specific logic I think that's a reasonable reason to look at refactors.

I just wanted to make sure our primary concerns don't center on problems we are unlikely to have.

File: lib/bolt_server/transport_app.rb [hunk 6]
```diff
@@ -630,6 +634,28 @@ def make_winrm_target(target_hash)
...
+      inventory = Bolt::Inventory.from_config(bolt_config, plugin)
+      target_list = inventory.get_targets('all').map do |target|
+        target_data = {
+          'name' => target.name,
+          'transport' => target.transport
+        }
+        target_data.merge(target.to_h)
+      end
+      target_list.to_json
+    rescue Bolt::Error => e
```
[the remark this answers is not in the document]
Comment (donoghuc) [reply lib/bolt_server/transport_app.rb:655]: I thought the opposite here, that we would want to broadly catch any error bolt might throw loading inventory.

[the remark this answers is not in the document]
Comment (ekinanp) [reply lib/bolt_server/transport_app.rb:655]: You're right, we should. I'm just not sure if those errors should return a `400` status because `400 Bad Request` usually means that something's up with the request. Here, the request is fine, just something went wrong w/ the parsing.

We had a similar discussion about this re: what status to return for an unsupported transport on the Orch-side, and we agreed with 422 there. Feels like this should also return a 422.

File: lib/bolt_server/transport_app.rb [hunk 7]
```diff
@@ -630,6 +642,23 @@ def make_winrm_target(target_hash)
       [400, e.message]
     end
 
+    # Returns a list of targets parsed from a Project inventory
+    #
+    # @param project_ref [String] the project_ref to compute the inventory from
+    get '/project_inventory_targets' do
+      bolt_config = config_from_project(params['project_ref'])
+      if bolt_config.inventoryfile && bolt_config.project.inventory_file.to_s != bolt_config.inventoryfile
+        raise Bolt::ValidationError, "Project inventory must be defined in #{bolt_config.project.inventory_file}"
```
Comment (ekinanp) [inline lib/bolt_server/transport_app.rb:651]: So this error will give me `<project_ref>/inventory.yaml` file, which could be different from how the user's storing their project because they don't know about `<project_ref>`.

Based on https://puppet.com/docs/bolt/latest/bolt_project_reference.html#inventoryfile, should this error change `#{bolt_config.project.inventory_file}` to `project/inventory.yaml` ?

File: lib/bolt_server/transport_app.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (ekinanp) [reply lib/bolt_server/transport_app.rb:658]: Might be worth bringing up ^ with the team though because an "invalidly structured inventory file" could warrant a different status code than a "valid Bolt inventory file but an invalid Orch inventory file" (which is what led to Orch's 422).

File: lib/bolt_server/transport_app.rb (hunk 7 above)
Comment (donoghuc) [reply lib/bolt_server/transport_app.rb:651]: oh yeah, good point.

File: lib/bolt_server/transport_app.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (nicklewis) [reply lib/bolt_server/transport_app.rb:659]: I think 500 for everything makes sense for now. We might later want to split it into 500 vs. 501 depending on whether the inventory file is using features that bolt-server doesn't support (I think that would just be plugins?).

File: spec/bolt_server/transport_app_spec.rb [hunk 8]
```diff
@@ -787,6 +795,80 @@ def post_over_transport(transport, action, body_content, multiple: false)
...
+            expect(target_list.first['protocol']).to eq('ssh')
+          end
+        end
+      end
+
+      context 'when inventory cannot be processed' do
+        let(:targets) {
+          [{ 'oops' => 'bad target' }]
+        }
+        it 'sets transport independent from protocol' do
```
Comment (ekinanp) [inline spec/bolt_server/transport_app_spec.rb:843]: Copy-paste error?

Comment (donoghuc) [reply spec/bolt_server/transport_app_spec.rb]: yep, updated

**Review by ekinanp — APPROVED** [review]

**Review by barriserloth — APPROVED** [review]

File: lib/bolt_server/transport_app.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (ekinanp) [reply lib/bolt_server/transport_app.rb:655]: Nit: `plugin` => `plugins` to keep it consistent with https://github.com/puppetlabs/bolt/blob/main/lib/bolt/cli.rb#L332

**Review by nicklewis — APPROVED** [review]

**Review by ekinanp — APPROVED** [review]
